src/freertos_drivers/ti/TivaSPI.hxx

```diff
@@ -0,0 +1,320 @@
+                {
+                    if (msg->rx_buf)
+                    {
+                        *rx_buf++ = data;
+                    }
+                    --rx_len;
+                }
+            }
+        }
+        while (tx_len || rx_len);
+
+        return msg->len;
+    }
+
+    /** Method to transmit/receive the data.
+     * @param msg message to transact.
+     */
+    __attribute__((optimize("-O3")))
+    int transfer_polled(struct spi_ioc_transfer *msg) override
+    {
+        /* set instance specific configuration */
+        set_configuration();
+
+        switch (bitsPerWord)
+        {
+            default:
+            case 8:
+                return transfer_polled<uint8_t>(msg);
+            case 16:
+                return transfer_polled<uint16_t>(msg);
+            case 32:
+                return transfer_polled<uint32_t>(msg);
+        }
+    }
+
+    /** Configure a DMA transaction.
+     * @param msg message to transact.
+     */
+    void config_dma(struct spi_ioc_transfer *msg);
+
+    /** Receives a word from the specified port.  This function gets a SPI word
+     * from the receive FIFO for the specified port.
+     * @param data is pointer to receive data variable.
+     * @return Returns the number of elements read from the receive FIFO.
+     */
+    __attribute__((optimize("-O3")))
+    long data_get_non_blocking(uint16_t *data)
+    {
+        if(HWREG(base_ + SSI_O_SR) & SSI_SR_RNE)
+        {
+            *data = HWREG(base_ + SSI_O_DR);
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /** Transmits a word on the specified port.  This function transmits a SPI
+     * word on the transmit FIFO for the specified port.
+     * @param data is data to be transmitted.
+     * @return Returns the number of elements written to the transmit FIFO.
+     */
+    __attribute__((optimize("-O3")))
+    long data_put_non_blocking(uint16_t data)
+    {
+        if(HWREG(base_ + SSI_O_SR) & SSI_SR_TNF)
+        {
+            HWREG(base_ + SSI_O_DR) = data;
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /** Waits until the word is transmitted on the specified port.  This
+     * function transmits a SPI word on the transmit FIFO for the specified
+     * port. This function waits until the space is available on transmit FIFO.
+     * @param data is data to be transmitted.
+     */
+    void data_put(uint16_t data)
+    {
+        while((HWREG(base_ + SSI_O_SR) & SSI_SR_TNF) == 0);
+        HWREG(base_ + SSI_O_DR) = data;
+    }
+
+    /** Set the instance local configuration.
+     */
+    void set_configuration()
+    {
+        // Disable port first for reconfiguration.
+        HWREG(base_ + SSI_O_CR1) &= ~SSI_CR1_SSE;
+        
+        HWREG(base_ + SSI_O_CR0) = spiCr0_;
+        HWREG(base_ + SSI_O_CPSR) = spiPrescaler_;
+        HWREG(base_ + SSI_O_CR1) = spiCr1_;
+    }
+
+    OSSem *sem_; /**< reference to the semaphore belonging to this bus */
```

Comment:
Does this get used anywhere? It may be dead code. Should probably check the CC3220 SPI driver too.

Reply:
I believe it is used to signal the DMA complete interrupt.

Reply:
Which explains why it seems unused, because DMA is still a TODO.